bob-access/src/credentials.rs

```diff
@@ -59,66 +74,223 @@ impl CredentialsKind {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Token(token.into()))
+    }
+
+    pub fn userpass_builder(username: impl Into<String>, password: impl Into<String>,) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Basic {
+            username: username.into(),
+            password: password.into(),
+        })
+    }
+
+    pub fn ip(&self) -> &Vec<SocketAddr> {
+        &self.address
+    }
+
+    pub fn replace_addresses(&mut self, addresses: Vec<SocketAddr>) {
+        self.address = addresses;
+    }
+
+    pub fn hostname(&self) -> &Option<String> {
+        &self.hostname
+    }
+
+    pub fn validate_internode(&self) -> bool {
+        (!self.address.is_empty() || self.hostname.is_some()) &&
+        self.kind.is_internode()
+    }
+
+    pub fn kind(&self) -> &CredentialsKind {
+        &self.kind
+    }
+
+    pub fn is_created_with_address(&self) -> bool {
+        self.hostname.is_none() && !self.address.is_empty()
+    }
+}
+
+#[derive(Debug)]
+pub struct DeclaredCredentialsBuilder {
+    kind: CredentialsKind,
+    address: Vec<SocketAddr>,
+    hostname: Option<String>,
+}
+
+impl DeclaredCredentialsBuilder {
+    fn new(kind: CredentialsKind) -> Self {
+        Self {
+            kind,
+            address: Vec::new(),
+            hostname: None,
+        }
+    }
+
+    pub fn with_address(mut self, address: SocketAddr) -> Self {
+        self.address = vec![address];
+        self
+    }
+
+    pub fn with_hostname(mut self, hostname: String) -> Self {
+        self.hostname = Some(hostname);
+        self
+    }
+
+    pub fn build(self) -> DeclaredCredentials {
+        DeclaredCredentials {
+            address: self.address,
+            kind: self.kind,
+            hostname: self.hostname,
+        }
+    }
+}
+
+
+#[derive(Debug)]
+pub struct DCredentialsResolveGuard {
+    credentials: DeclaredCredentials,
+    resolve_state: ResolveState,
+    resolve_threshold: Instant,
+    resolve_protection_period_ms: u64,
+}
+
+impl DCredentialsResolveGuard {
+    pub fn new(credentials: DeclaredCredentials, resolve_protection_period_ms: u64) -> Self {
+        Self {
+            credentials,
+            resolve_state: ResolveState::Resolved(0),
+            resolve_threshold: Instant::now() + Duration::from_millis(resolve_protection_period_ms),
+            resolve_protection_period_ms,
+        }
+    }
+
+    pub fn update_resolve_state(&mut self, authenticated: bool) -> bool {
+        if self.credentials.is_created_with_address() {
+            return false;
+        }
+        
+        if authenticated {
+            self.resolve_state.update(authenticated)
+        } else {
+            if Instant::now() >= self.resolve_threshold {
+                self.resolve_state.update(authenticated)
+            } else {
+                false
+            }
+        }
+    }
+
+    pub fn needs_update(&self, authenticated: bool) -> bool {
```

Comment:
Add comment that the logic of this function is tightly coupled with `update_resolve_state`